### PR TITLE
Fix the broken Dockerfile released in v3.3.0

### LINKED ISSUE #2141: Docker image throwing unknown module exception
### Issue workflow progress

_Progress of the issue based on the [Contributor Workflow](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md#a-typical-contributor-workflow)_

- [ ] 1. The issue provides a reproduction available on GitHub, Stackblitz or CodeSandbox
> Make sure to fork this template and run `yarn generate` in the terminal.
>
> Please make sure the Codegen and plugins version under `package.json` matches yours.
- [ ] 2. A failing test has been provided
- [ ] 3. A local solution has been provided
- [ ] 4. A pull request is pending review

---

**Describe the bug**

Running the following command with the latest docker image version

```
graphql-inspector diff https://app.launchnotes.io/graphql frontend/src/data/schema.graphql
```

Results in the following exception:

```
node:internal/modules/cjs/loader:936
  throw err;
  ^
Error: Cannot find module 'tslib'
Require stack:
- /usr/local/share/graphql-inspector/cli/dist/index.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:[27](https://gitlab.com/shiny-planes/the-goat/-/jobs/2729717131#L27))
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/usr/local/share/graphql-inspector/cli/dist/index.js:6:15)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:[32](https://gitlab.com/shiny-planes/the-goat/-/jobs/2729717131#L32))
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/usr/local/share/graphql-inspector/cli/dist/index.js' ]
}
```

**To Reproduce**

```
graphql-inspector diff https://app.launchnotes.io/graphql frontend/src/data/schema.graphql
```

**Expected behavior**

I expect the diff to run successfully.

**Environment:**

- Latest docker image

**Additional context**

We have been running this in CI for about six months with no issue until the recent release pushed a new version to docker.  Since we're using the latest tag, builds after the release began to fail.

### PR DESCRIPTION
**Pull request by ayumitamai97**
## Description

Fixes #2141

I fixed the broken Dockerfile released in v3.3.0

Changes:

- Copy `package.json` from build stage to dist stage
- Run `yarn install` in `NODE_ENV=production` so that the image can include the dependencies (excluding devDependencies)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Screenshots/Sandbox (if appropriate/relevant):

DockerHub:
https://hub.docker.com/layers/graphql-inspector/ayumitamai97/graphql-inspector/3.3.0-modified/images/sha256-2abe9cd96a5bf6d74f173ed643c0d9821b521a5cf3c55b7fc4eb1b71902851e2?context=explore

How to test:

```
docker run --pull=always --rm -v $PWD:/app \
  ayumitamai97/graphql-inspector:3.3.0-modified \
  graphql-inspector introspect test.graphql
```

スクリーンショット 2022-07-24 4 26 49

## How Has This Been Tested?

- [x] Docker image can built successfully
- [x] The command `docker run graphql-inspector ...` like [this](https://github.com/kamilkisiela/graphql-inspector/blob/ef23d6c9f2748deae9b64859841a8cbdbfa1590a/website/docs/introduction/installation.mdx#docker) runs successfully

**Test Environment**:

- OS: MacOS Monterey v12.3.1 
- Docker Desktop for Mac
- Using Docker BuildKit

## Checklist:

- [x] I have followed the [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

None

### CODE REVIEW

File: Dockerfile
Comment (ayumitamai97) [inline Dockerfile:9]: I tried to reduce the build duration by this:

```Dockerfile
COPY ./package.json ./
RUN yarn install

COPY . .
RUN yarn build
```

([ref](https://stackoverflow.com/a/67333861))

but it did not work since this is monorepo and the submodules has many `package.json`s 🥲

File: Dockerfile
Comment (ayumitamai97) [inline Dockerfile:20]: The cause of the error reported in #2141 is that the final Docker image did not have required `node_modules` installed.

File: Dockerfile
Comment (spawnia) [inline Dockerfile:14]: Better use cd in the RUN instruction below to reduce the number of layers

File: Dockerfile
Comment (spawnia) [inline Dockerfile:22]: Better combine all RUN statements to reduce layers

File: Dockerfile
Comment (spawnia) [inline Dockerfile:26]: This should move back to where it was, as it changes seldom and thus reduces image churn when put in an earlier layer

File: Dockerfile
Comment (spawnia) [inline Dockerfile:11]: Why not stick with alpine?

**Comment by ayumitamai97** [comment]
Thank you for the review, I'm fixing :bow:

File: Dockerfile
Comment (ayumitamai97) [reply Dockerfile:11]: I thought `slim` is smaller, but it was a mistake

**Comment by ayumitamai97** [comment]
I fixed Dockerfile following @ spawnia's instruction. Is my understanding right? If not, could you please suggest changes? Thank you!

**Review by spawnia — APPROVED** [review]
Nicely done. This approach makes sense to me.

**Comment by jillesme** [comment]
@ayumitamai97 ready to merge soon?

**Comment by ayumitamai97** [comment]
@jillesme I can't merge this PR until @kamilkisiela's approval. I'm waiting for it but he seems busy

**Comment by jillesme** [comment]
@kamilkisiela would you be able to approve this? We currently have to disable GraphQL Inspector because the image won't build.

CleanShot 2022-08-08 at 13 44 12@2x

**Comment by ayumitamai97** [comment]
@kamilkisiela I resolved conflict with #2150. Please review this Pull Request, and if it's ok, could you please merge this?

Some people including me are waiting for the problem to be fixed.

**Comment by doriandekoning** [comment]
We're still waiting on this, are there any updates or an ETA?

**Comment by kamilkisiela** [comment]
Thank you <3

**Pull request merged**
